File: supabase/functions/telegram-bot/ascii_converter.ts
import { Image, createCanvas } from "https://deno.land/x/canvas@v1.4.1/mod.ts"

interface ASCIISettings {
  width: number;
  contrast: number;
  brightness: number;
  color_mode: string;
}

export class ASCIIArtConverter {
  private readonly ASCII_CHARS = '$@B%8&WM#*oahkbdpqwmZO0QLCJUYXzcvunxrjft/\\|()1{}[]?-_+~<>i!lI;:,"^`\'. '
  private readonly settings: ASCIISettings
  private readonly fontSize: number = 20

  constructor(settings: ASCIISettings) {
    this.settings = settings
  }

  private adjustImage(imageData: ImageData): ImageData {
    const data = new Uint8ClampedArray(imageData.data)
    const mean = this.calculateMean(data)

    for (let i = 0; i < data.length; i += 4) {
      // Apply contrast
      for (let j = 0; j < 3; j++) {
        data[i + j] = (data[i + j] - mean) * this.settings.contrast + mean
      }

      // Apply brightness
      for (let j = 0; j < 3; j++) {
        data[i + j] = data[i + j] * this.settings.brightness
      }
    }

    return new ImageData(data, imageData.width, imageData.height)
  }

  private calculateMean(data: Uint8ClampedArray): number {
    let sum = 0
    for (let i = 0; i < data.length; i += 4) {
      sum += (data[i] + data[i + 1] + data[i + 2]) / 3
    }
    return sum / (data.length / 4)
  }

  private getColor(r: number, g: number, b: number): [number, number, number] {
    const avg = (r + g + b) / 3

    switch (this.settings.color_mode) {
      case 'true_color':
        return [r, g, b]
      case 'mono':
        return avg < 128 ? [0, 0, 0] : [255, 255, 255]
      case 'green':
        return [0, avg, 0]
      case 'blue':
        return [0, 0, avg]
      case 'red':
        return [avg, 0, 0]
      case 'cyan':
        return [0, avg, avg]
      case 'magenta':
        return [avg, 0, avg]
      case 'yellow':
        return [avg, avg, 0]
      default: // grayscale
        return [avg, avg, avg]
    }
  }

  private resizeImage(image: Image): Image {
    const origAspect = image.height / image.width
    const targetWidth = this.settings.width
    const targetHeight = Math.floor(targetWidth * origAspect)

    const canvas = createCanvas(targetWidth, targetHeight)
    const ctx = canvas.getContext('2d')
    ctx.drawImage(image, 0, 0, targetWidth, targetHeight)

    return canvas
  }

  private toGrayscale(imageData: ImageData): ImageData {
    const data = new Uint8ClampedArray(imageData.data)
    for (let i = 0; i < data.length; i += 4) {
      const avg = (data[i] + data[i + 1] + data[i + 2]) / 3
      data[i] = data[i + 1] = data[i + 2] = avg
    }
    return new ImageData(data, imageData.width, imageData.height)
  }

  private pixelsToAscii(imageData: ImageData): string {
    const data = imageData.data
    const width = imageData.width
    const height = imageData.height
    let ascii = ''

    for (let y = 0; y < height; y++) {
      for (let x = 0; x < width; x++) {
        const idx = (y * width + x) * 4
        const brightness = data[idx]
        const charIndex = Math.floor(brightness / 255 * (this.ASCII_CHARS.length - 1))
        ascii += this.ASCII_CHARS[charIndex]
      }
      ascii += '\n'
    }

    return ascii
  }

  private createAsciiImage(ascii: string, originalImage: Image): Image {
    const lines = ascii.split('\n')
    const charWidth = 10 // Approximate width of a character
    const charHeight = 20 // Approximate height of a character

    const imgWidth = charWidth * lines[0].length
    const imgHeight = Math.floor(imgWidth * (originalImage.height / originalImage.width))

    const canvas = createCanvas(imgWidth, imgHeight)
    const ctx = canvas.getContext('2d')

    // Set dark background
    ctx.fillStyle = 'rgb(32, 32, 32)'
    ctx.fillRect(0, 0, imgWidth, imgHeight)

    // Calculate spacing
    const totalTextHeight = charHeight * lines.length
    const scaleFactor = imgHeight / totalTextHeight
    const adjustedCharHeight = charHeight * scaleFactor

    // Draw characters
    ctx.font = `${this.fontSize}px monospace`
    ctx.textBaseline = 'top'

    for (let y = 0; y < lines.length; y++) {
      const line = lines[y]
      for (let x = 0; x < line.length; x++) {
        const char = line[x]
        const posX = x * charWidth
        const posY = y * adjustedCharHeight

        // Get color from original image
        const origX = Math.floor(x * originalImage.width / line.length)
        const origY = Math.floor(y * originalImage.height / lines.length)
        const pixelData = ctx.getImageData(origX, origY, 1, 1).data
        const [r, g, b] = this.getColor(pixelData[0], pixelData[1], pixelData[2])

        ctx.fillStyle = `rgb(${r}, ${g}, ${b})`
        ctx.fillText(char, posX, posY)
      }
    }

    return canvas
  }

  public async convert(imageData: ArrayBuffer): Promise<ArrayBuffer> {
    // Create image from buffer
    const image = await createImageBitmap(new Blob([imageData]))
    
    // Resize image
    const resized = this.resizeImage(image)
    
    // Get image data and apply adjustments
    const ctx = resized.getContext('2d')
    let data = ctx.getImageData(0, 0, resized.width, resized.height)
    data = this.adjustImage(data)
    
    // Convert to grayscale for ASCII conversion
    const gray = this.toGrayscale(data)
    
    // Convert to ASCII
    const ascii = this.pixelsToAscii(gray)
    
    // Create final image
    const asciiImage = this.createAsciiImage(ascii, image)
    
    // Convert to PNG buffer
    return await asciiImage.encode('png')
  }
} 

File: ascii_converter.py
import numpy as np
from PIL import Image, ImageDraw, ImageFont
import os

class ASCIIArtConverter:
    # ASCII characters from darkest to lightest (expanded set)
    ASCII_CHARS = r'$@B%8&WM#*oahkbdpqwmZO0QLCJUYXzcvunxrjft/\|()1{}[]?-_+~<>i!lI;:,"^`\'. '
    
    def __init__(self, width=100, contrast=1.0, brightness=1.0, color_mode='true_color'):
        self.width = width
        self.contrast = contrast
        self.brightness = brightness
        self.color_mode = color_mode
        # Fixed font size for consistency
        self.font_size = 20
        # Try to load a monospace font
        try:
            self.font = ImageFont.truetype("Courier New", self.font_size)
        except:
            try:
                self.font = ImageFont.truetype("DejaVu Sans Mono", self.font_size)
            except:
                try:
                    self.font = ImageFont.truetype("Consolas", self.font_size)
                except:
                    self.font = ImageFont.load_default()
    
    def _adjust_image(self, image):
        """Apply contrast and brightness adjustments."""
        # Convert to float for calculations
        img_array = np.array(image, dtype=float)
        
        # Apply contrast
        mean = np.mean(img_array)
        img_array = (img_array - mean) * self.contrast + mean
        
        # Apply brightness
        img_array = img_array * self.brightness
        
        # Clip values to valid range
        img_array = np.clip(img_array, 0, 255)
        
        # Convert back to uint8
        return Image.fromarray(np.uint8(img_array))
    
    def _get_color(self, color):
        """Get color based on color_mode."""
        if self.color_mode == 'true_color':
            return color
        elif self.color_mode == 'mono':
            # Black and white only
            avg = sum(color) / 3
            return (0, 0, 0) if avg < 128 else (255, 255, 255)
        elif self.color_mode == 'green':
            avg = sum(color) / 3
            return (0, int(avg), 0)
        elif self.color_mode == 'blue':
            avg = sum(color) / 3
            return (0, 0, int(avg))
        elif self.color_mode == 'red':
            avg = sum(color) / 3
            return (int(avg), 0, 0)
        elif self.color_mode == 'cyan':
            avg = sum(color) / 3
            return (0, int(avg), int(avg))
        elif self.color_mode == 'magenta':
            avg = sum(color) / 3
            return (int(avg), 0, int(avg))
        elif self.color_mode == 'yellow':
            avg = sum(color) / 3
            return (int(avg), int(avg), 0)
        else:  # grayscale
            avg = int(sum(color) / 3)
            return (avg, avg, avg)
    
    def _resize_image(self, image):
        """Resize image maintaining aspect ratio."""
        # Get original dimensions
        orig_width, orig_height = image.size
        orig_aspect = orig_height / orig_width
        
        # Calculate target dimensions
        target_width = self.width
        target_height = int(target_width * orig_aspect)
        
        return image.resize((target_width, target_height))
    
    def _to_grayscale(self, image):
        """Convert image to grayscale."""
        return image.convert('L')
    
    def _pixels_to_ascii(self, image):
        """Convert pixels to ASCII characters."""
        pixels = np.array(image)
        ascii_str = []
        
        # Convert each pixel to an ASCII character
        for row in pixels:
            ascii_str.append(''.join([self.ASCII_CHARS[int(pixel / 255 * (len(self.ASCII_CHARS) - 1))] 
                                    for pixel in row]))
        return '\n'.join(ascii_str)
    
    def _create_ascii_image(self, ascii_str, original_image):
        """Create an image from ASCII string with color based on original image."""
        # Calculate dimensions
        lines = ascii_str.split('\n')
        char_width = self.font.getbbox('A')[2]
        char_height = self.font.getbbox('A')[3]
        
        # Get original dimensions for aspect ratio
        orig_width, orig_height = original_image.size
        orig_aspect = orig_height / orig_width
        
        # Calculate image dimensions to match original aspect ratio
        img_width = char_width * len(lines[0])
        img_height = int(img_width * orig_aspect)
        
        # Create new image with dark background
        ascii_image = Image.new('RGB', (img_width, img_height), (32, 32, 32))
        draw = ImageDraw.Draw(ascii_image)
        
        # Get color information from original image
        original_colors = np.array(original_image.resize((len(lines[0]), len(lines))))
        
        # Calculate vertical spacing to distribute lines evenly
        total_text_height = char_height * len(lines)
        scale_factor = img_height / total_text_height
        adjusted_char_height = char_height * scale_factor
        
        # Draw each character with corresponding color
        for y, line in enumerate(lines):
            for x, char in enumerate(line):
                # Get color from original image and apply color mode
                color = tuple(map(int, original_colors[y, x]))
                final_color = self._get_color(color)
                # Calculate position to maintain aspect ratio
                pos_x = x * char_width
                pos_y = y * adjusted_char_height
                # Draw character
                draw.text((pos_x, pos_y), char, font=self.font, fill=final_color)
        
        return ascii_image
    
    def convert(self, image_input, output_dir="ascii_output"):
        """Convert image to ASCII art and save both text and image versions."""
        os.makedirs(output_dir, exist_ok=True)
        
        # Handle both file paths and PIL Image objects
        if isinstance(image_input, str):
            # It's a file path
            with Image.open(image_input) as img:
                if img.mode != 'RGB':
                    img = img.convert('RGB')
                img = self._resize_image(img)
                img = self._adjust_image(img)
                original_resized = img.copy()
                img = self._to_grayscale(img)
                ascii_str = self._pixels_to_ascii(img)
                ascii_image = self._create_ascii_image(ascii_str, original_resized)
        else:
            # It's a PIL Image object
            img = image_input
            if img.mode != 'RGB':
                img = img.convert('RGB')
            img = self._resize_image(img)
            img = self._adjust_image(img)
            original_resized = img.copy()
            img = self._to_grayscale(img)
            ascii_str = self._pixels_to_ascii(img)
            ascii_image = self._create_ascii_image(ascii_str, original_resized)
        
        return ascii_str, ascii_image

if __name__ == "__main__":
    # Example usage
    converter = ASCIIArtConverter(font_size=12, width=100)
    if os.path.exists("test.jpg"):
        ascii_str, ascii_image = converter.convert("test.jpg")
        print(ascii_str)